File: Graph/csharp/UndirectedPath.cs
using System.Collections.Generic;

namespace GraphSolutions;

public static class UndirectedPath
{
    public static bool Solve(IEnumerable<(string, string)> edges, string src, string dst)
    {
        var graph = GraphUtils.BuildUndirectedGraph(edges);
        var visited = new HashSet<string>();
        return Dfs(graph, src, dst, visited);
    }

    private static bool Dfs(Dictionary<string, List<string>> graph, string node, string dst, HashSet<string> visited)
    {
        if (node == dst)
        {
            return true;
        }

        if (!visited.Add(node))
        {
            return false;
        }

        if (!graph.TryGetValue(node, out var neighbors))
        {
            return false;
        }

        foreach (var neighbor in neighbors)
        {
            if (Dfs(graph, neighbor, dst, visited))
            {
                return true;
            }
        }

        return false;
    }
}

File: Graph/python/undirectedPath.py
from typing import List, Tuple

from .graph_utils import Graph, build_undirected_graph


def undirected_path(edges: List[Tuple[str, str]], src: str, dst: str) -> bool:
    graph = build_undirected_graph(edges)
    return _dfs(graph, src, dst, set())


def _dfs(graph: Graph, node: str, dst: str, visited: set[str]) -> bool:
    if node == dst:
        return True
    if node in visited:
        return False

    visited.add(node)
    for neighbor in graph.get(node, []):
        if _dfs(graph, neighbor, dst, visited):
            return True
    return False
